File: dataflow_bench_in_cpp/join_op.cpp
// Copyright 2016 IDEXX
#include "dataflow_bench_in_cpp/join_op.h"
#include <cassert>
#include <set>

// Constructor
JoinOp::JoinOp(OpInterface* next) :
  table_(),
  next_(next) {
  assert(next);
}

void JoinOp::Add(int to_add) {
  table_.insert(to_add);
}

void JoinOp::Invoke(int incoming) {
  if (table_.count(incoming) == 0) {
    next_->Invoke(incoming);
  }
}

File: dataflow_bench_in_java/JoinOp.java
import java.util.Set;
import java.util.HashSet;

public class JoinOp implements OpInterface
{
    private OpInterface next;
    private Set<Integer> table;
    
    // Constructor
    public JoinOp(OpInterface next)
    {
        this.next = next;
        this.table = new HashSet<Integer>();
    }

    // Mutator
    public void add(int to_add)
    {
        this.table.add(to_add);
    }
    
    @Override
    public void invoke(int incoming)
    {
        if (!this.table.contains(incoming)) {
            next.invoke(incoming);
        }
    }
}
